couchdb/server/compiler.py

```diff
@@ -297,3 +319,69 @@ def require(path, module=None):
                 _visited_ids.pop()
 
     return require
+
+
+def compile_func(funsrc, ddoc=None, context=None, encoding='utf-8', **options):
+    """Compile source code and extract function object from it.
+
+    :param funsrc: Python source code.
+    :type funsrc: unicode
+
+    :param ddoc: Optional argument which must represent design document.
+    :type ddoc: dict
+
+    :param context: Custom context objects which function could operate with.
+    :type context: dict
+
+    :param options: Compiler config options.
+
+    :param encoding: Encoding of source code.
+    :type encoding: str
+
+    :return: Function object.
+
+    :raises:
+        - :exc:`~couchdb.server.exceptions.Error`
+          If source code compilation failed or it doesn't contains function
+          definition.
+
+    .. note::
+        ``funsrc`` should contains only one function definition and import
+        statements (optional) or :exc:`~couchdb.server.exceptions.Error`
+        will be raised.
+
+    """
+    if not context:
+        context = DEFAULT_CONTEXT.copy()
+    else:
+        context, _ = DEFAULT_CONTEXT.copy(), context
+        context.update(_)
+    if ddoc is not None:
+        context['require'] = require(ddoc, context, **options)
+
+    globals_ = {}
+    try:
+        bytecode = compile_to_bytecode(funsrc, encoding=encoding)
+        exec(bytecode, context, globals_)
+    except Exception as err:
+        log.exception('Failed to compile source code:\n%s', funsrc)
+        raise Error('compilation_error', str(err))
+
+    msg = None
+    func = None
+    for item in globals_.values():
+        if isinstance(item, FunctionType):
+            if func is None:
+                func = item
+            else:
+                msg = 'Multiple functions are defined. Only one is allowed.'
+        elif not isinstance(item, ModuleType):
+            msg = 'Only functions could be defined at top level namespace'
```

Comment:
If we throw all the function and module(s) into global scope, it will be handy for user. But do we introduce some terrible side effect?

```diff
diff --git a/couchdb/server/compiler.py b/couchdb/server/compiler.py
index c91a4ef..3ba2022 100644
--- a/couchdb/server/compiler.py
+++ b/couchdb/server/compiler.py
@@ -385,4 +385,8 @@ def compile_func(funsrc, ddoc=None, context=None, encoding='utf-8', **options):
     if msg is not None:
         log.error('%s\n%s', msg, funsrc)
         raise Error('compilation_error', msg)
+
+    context.update(globals_)
+    globals_.clear()
+
     return func
diff --git a/couchdb/tests/server/compiler.py b/couchdb/tests/server/compiler.py
index ac50b3b..edc0c32 100644
--- a/couchdb/tests/server/compiler.py
+++ b/couchdb/tests/server/compiler.py
@@ -254,19 +259,23 @@ class CompilerTestCase(unittest.TestCase):
     def test_allow_imports(self):
         funsrc = (
             'import math\n'
-            'def test(math=math):\n'
+            'def test():\n'
             '  return math.sqrt(42*42)'
         )
         func = compiler.compile_func(funsrc)
         self.assertEqual(func(), 42)

-    def test_fail_for_non_clojured_imports(self):
+    def test_for_module_scope(self):
         funsrc = (
             'import math\n'
             'def test():\n'
-            '  return math.sqrt(42*42)')
+            '  assert "math" in globals()\n'
+            '  return math.sqrt(42*42)\n'
+            # At the time being executed,
+            # we do not move module/func into globals yet
+            'assert "math" not in globals()\n'
+        )
         func = compiler.compile_func(funsrc)
-        self.assertRaises(NameError, func)
+        self.assertEqual(func(), 42)

     def test_ascii_function_source_string(self):
         funsrc = 'def test(): return 42'
```